pyocd/target/builtin/target_ytm32b1md0.py

```diff
@@ -0,0 +1,98 @@
+#
+#     http://www.apache.org/licenses/LICENSE-2.0
+#
+# Unless required by applicable law or agreed to in writing, software
+# distributed under the License is distributed on an "AS IS" BASIS,
+# WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+# See the License for the specific language governing permissions and
+# limitations under the License.
+
+from ...coresight.coresight_target import CoreSightTarget
+from ...coresight.cortex_m import CortexM 
+from ...core.memory_map import (FlashRegion, RamRegion, MemoryMap)
+
+
+FLASH_ALGO = {
+    'load_address' : 0x20000000,
+
+    # Flash algorithm as a hex string
+    'instructions': [
+    0xe00abe00,
+    0xb5704770, 0x460d4604, 0x48344616, 0x60084934, 0x60084834, 0x68404608, 0x00400840, 0x48326048,
+    0x21206900, 0x49304308, 0x20036108, 0xbf006008, 0x6880482d, 0x0f800780, 0xd1f92803, 0x492a482b,
+    0x20006048, 0xbf006008, 0x68804827, 0x0f800780, 0xd1f92800, 0xf836f000, 0x4601bd70, 0x47702000,
+    0x2006b510, 0x72084922, 0xf82cf000, 0xb510bd10, 0xf7ff4604, 0x481fffc5, 0x20046020, 0x7208491c,
+    0xf820f000, 0xb5f7bd10, 0x4607b082, 0x9c04460e, 0x2008463d, 0xbf009001, 0xc501cc01, 0xc501cc01,
+    0x49132002, 0xf0007208, 0x2800f80d, 0x2001d002, 0xbdf0b005, 0x1a369801, 0xffa2f7ff, 0xd1eb2e00,
+    0xe7f52000, 0x480abf00, 0x21807900, 0x28004008, 0x4807d0f9, 0x210c7900, 0x47704008, 0x0000b631,
+    0x4006a000, 0x0000c278, 0x4007c000, 0x00010001, 0x40010000, 0x12345678, 0x00000000
+    ],
+
+    # Relative function addresses
+    'pc_init': 0x20000007,
+    'pc_unInit': 0x2000005f,
+    'pc_program_page': 0x2000008b,
+    'pc_erase_sector': 0x20000073,
+    'pc_eraseAll': 0x20000065,
+
+    'static_base' : 0x20000000 + 0x00000004 + 0x000000f8,
+    'begin_stack' : 0x20000300,
+    'begin_data' : 0x20000000 + 0x1000,
+    'page_size' : 0x80,
+    'analyzer_supported' : False,
+    'analyzer_address' : 0x00000000,
+    'page_buffers' : [0x20001000, 0x20001080],   # Enable double buffering
+    'min_program_length' : 0x80,
+
+    # Relative region addresses and sizes
+    'ro_start': 0x0,
+    'ro_size': 0xf8,
+    'rw_start': 0xf8,
+    'rw_size': 0x4,
+    'zi_start': 0xfc,
+    'zi_size': 0x0,
+
+    # Flash information
+    'flash_start': 0x0,
+    'flash_size': 0x80000,
+    'sector_sizes': (
+        (0x0, 0x800),
+    )
+}
+
+class YTM32B1MD0(CoreSightTarget):
+
+    VENDOR = "YTMicro"
+    
+    MEMORY_MAP = MemoryMap(
+        FlashRegion(    start=0x0000,           length=0x80000,      blocksize=0x800, is_boot_memory=True,
+            algo=FLASH_ALGO),
+        RamRegion(      start=0x20000000,  length=0x8000)
+        )
+
+    def __init__(self, session):
+        super(YTM32B1MD0, self).__init__(session, self.MEMORY_MAP)
+
+    def create_init_sequence(self):
+        seq = super(YTM32B1MD0, self).create_init_sequence()
+        seq.wrap_task('discovery',
+            lambda seq: seq.replace_task('create_cores', self.create_cores)
+            )
```

Comment:
Is there a reason you need to override the create_cores task? I'm just curious why standard CoreSight discovery would not work. Please add a comment here describing the situation so other people looking at this code (like me in a few months or years! 😄 ) will understand.

Reply:
I not sure why CoreSight can not find the core, we have two APs, core on AP0, but CoreSight seem searching core on AP1.

Reply:
What does the log output look like what you run `pyocd cmd -vv -c exit` without the modified tasks? This should work, so I'd like to understand if there is a bug in pyocd or some other change that I need to make.

Reply:
Log below is trying download elf to chip without change AP.
```
0001090:DEBUG:pemicro:Opened PEMicro library: /usr/local/lib/python3.9/site-packages/pypemicro/libs/MacOS/unitacmp-64.dylib
0001150:DEBUG:session:Project directory: /Users/major/work/yuntu/YTMicroSDK/demos/YTM32B1MD0/hello_world/GCC
0001150:INFO:board:Target type is ytm32b1md0
0001154:DEBUG:dap_access_cmsis_dap:CMSIS-DAP probe 0001A0000000: protocol version 2.0.0
0001159:DEBUG:sequencer:Running task load_svd
0001159:DEBUG:sequencer:Running task pre_connect
0001159:DEBUG:sequencer:Running task dp_init
0001159:DEBUG:sequencer:Running task lock_probe
0001159:DEBUG:sequencer:Running task get_probe_capabilities
0001159:DEBUG:sequencer:Running task connect
0001470:DEBUG:dap:Default wire protocol selected; using SWD
0001473:DEBUG:swj:Sending deprecated SWJ sequence to select SWD
0001485:INFO:dap:DP IDR = 0x6ba02477 (v2 rev6)
0001485:DEBUG:sequencer:Running task clear_sticky_err
0001486:DEBUG:sequencer:Running task power_up_debug
0001489:DEBUG:sequencer:Running task check_version
0001489:DEBUG:sequencer:Running task unlock_probe
0001489:DEBUG:sequencer:Running task create_discoverer
0001489:DEBUG:sequencer:Running task discovery
0001490:DEBUG:sequencer:Running task find_aps
0001503:DEBUG:sequencer:Running task create_aps
0001503:DEBUG:sequencer:Running task create_ap.0
0001506:INFO:ap:AHB-AP#0 IDR = 0x84770001 (AHB-AP var0 rev8)
0001511:DEBUG:ap:AHB-AP#0 default HPROT=3 HNONSEC=0
0001513:DEBUG:ap:AHB-AP#0 implemented HPROT=f HNONSEC=1
0001515:DEBUG:sequencer:Running task create_ap.1
0001517:INFO:ap:APB-AP#1 IDR = 0x54770002 (APB-AP var0 rev5)
0001525:DEBUG:ap:APB-AP#1 default HPROT=0 HNONSEC=0
0001528:DEBUG:ap:APB-AP#1 implemented HPROT=0 HNONSEC=0
0001531:DEBUG:sequencer:Running task find_components
0001531:DEBUG:sequencer:Running task create_cores
0001531:DEBUG:sequencer:Running task create_components
0001531:DEBUG:sequencer:Running task check_for_cores
0001531:DEBUG:session:uninit session <pyocd.core.session.Session object at 0x10f28cbe0>
0001535:DEBUG:hidapi_backend:closing interface
0001536:CRITICAL:__main__:No cores were discovered!
Traceback (most recent call last):
  File "/usr/local/lib/python3.9/site-packages/pyocd/__main__.py", line 150, in run
    status = cmd.invoke()
  File "/usr/local/lib/python3.9/site-packages/pyocd/subcommands/load_cmd.py", line 96, in invoke
    with session:
  File "/usr/local/lib/python3.9/site-packages/pyocd/core/session.py", line 341, in __enter__
    self.open()
  File "/usr/local/lib/python3.9/site-packages/pyocd/core/session.py", line 459, in open
    self._board.init()
  File "/usr/local/lib/python3.9/site-packages/pyocd/board/board.py", line 99, in init
    self.target.init()
  File "/usr/local/lib/python3.9/site-packages/pyocd/core/soc_target.py", line 117, in init
    seq.invoke()
  File "/usr/local/lib/python3.9/site-packages/pyocd/utility/sequencer.py", line 208, in invoke
    resultSequence = call()
  File "/usr/local/lib/python3.9/site-packages/pyocd/coresight/coresight_target.py", line 244, in check_for_cores
    raise exceptions.DebugError("No cores were discovered!")
pyocd.core.exceptions.DebugError: No cores were discovered!
```

Log below is download when assign core manually:

```
0001409:DEBUG:pemicro:Opened PEMicro library: /usr/local/lib/python3.9/site-packages/pypemicro/libs/MacOS/unitacmp-64.dylib
0001471:DEBUG:session:Project directory: /Users/major/work/yuntu/YTMicroSDK/demos/YTM32B1MD0/hello_world/GCC
0001473:INFO:board:Target type is ytm32b1md0
0001476:DEBUG:dap_access_cmsis_dap:CMSIS-DAP probe 0001A0000000: protocol version 2.0.0
0001481:DEBUG:sequencer:Running task load_svd
0001481:DEBUG:sequencer:Running task pre_connect
0001481:DEBUG:sequencer:Running task dp_init
0001481:DEBUG:sequencer:Running task lock_probe
0001481:DEBUG:sequencer:Running task get_probe_capabilities
0001481:DEBUG:sequencer:Running task connect
0001793:DEBUG:dap:Default wire protocol selected; using SWD
0001796:DEBUG:swj:Sending deprecated SWJ sequence to select SWD
0001808:INFO:dap:DP IDR = 0x6ba02477 (v2 rev6)
0001808:DEBUG:sequencer:Running task clear_sticky_err
0001808:DEBUG:sequencer:Running task power_up_debug
0001812:DEBUG:sequencer:Running task check_version
0001812:DEBUG:sequencer:Running task unlock_probe
0001812:DEBUG:sequencer:Running task create_discoverer
0001812:DEBUG:sequencer:Running task discovery
0001812:DEBUG:sequencer:Running task find_aps
0001823:DEBUG:sequencer:Running task create_aps
0001823:DEBUG:sequencer:Running task create_ap.0
0001826:INFO:ap:AHB-AP#0 IDR = 0x84770001 (AHB-AP var0 rev8)
0001830:DEBUG:ap:AHB-AP#0 default HPROT=3 HNONSEC=0
0001832:DEBUG:ap:AHB-AP#0 implemented HPROT=f HNONSEC=1
0001834:DEBUG:sequencer:Running task create_ap.1
0001836:INFO:ap:APB-AP#1 IDR = 0x54770002 (APB-AP var0 rev5)
0001842:DEBUG:ap:APB-AP#1 default HPROT=0 HNONSEC=0
0001844:DEBUG:ap:APB-AP#1 implemented HPROT=0 HNONSEC=0
0001846:DEBUG:sequencer:Running task find_components
0001847:DEBUG:sequencer:Running task create_cores
0001850:INFO:cortex_m:CPU core #0 is Cortex-M33 r0p4
0001858:INFO:cortex_m:FPU present: FPv5-SP-D16-M
0001859:DEBUG:sequencer:Running task create_components
0001859:DEBUG:sequencer:Running task check_for_cores
0001859:DEBUG:sequencer:Running task halt_on_connect
0001859:DEBUG:cortex_m:halting core 0
0001863:DEBUG:sequencer:Running task post_connect
0001864:DEBUG:sequencer:Running task post_connect_hook
0001864:DEBUG:sequencer:Running task create_flash
0001864:DEBUG:sequencer:Running task notify
0001865:INFO:load_cmd:Loading /Users/major/work/yuntu/YTMicroSDK/demos/YTM32B1MD0/hello_world/GCC/build/flash.elf
0001869:DEBUG:file_programmer:Writing segment LMA:0x00000000, VMA:0x00000000, size 820
0001870:DEBUG:file_programmer:Writing segment LMA:0x00000400, VMA:0x00000400, size 11452
0001870:DEBUG:file_programmer:Writing segment LMA:0x000030bc, VMA:0x1fff8400, size 56
0001870:DEBUG:file_programmer:Skipping segment LMA:0x1fff8000, VMA:0x1fff8000, size 0
0001870:DEBUG:cortex_m:halting core 0
0001872:DEBUG:cortex_m:set reset catch, core 0
0001873:DEBUG:cortex_m:halting core 0
0001878:DEBUG:cortex_m:reset, core 0, type=SW_SYSRESETREQ
0002018:DEBUG:cortex_m:clear reset catch, core 0
0002048:DEBUG:cortex_m:resuming core 0
0002049:DEBUG:manager:added=[] removed=[]
0002049:DEBUG:manager:bps after flush={}
0002160:DEBUG:builder:Analyze time: 0.000026
0002160:DEBUG:builder:Using double buffer sector erase program
[                                                  ]   0%0002160:DEBUG:cortex_m:halting core 0
0002177:DEBUG:cortex_m:resuming core 0
.......
```

Reply:
Thanks for the debug logs.

Do you know if AP#0 has a missing ROM table? Any AHB-AP that has an attached Cortex-M core is supposed to have a CoreSight ROM table.

Also, what is the APB-AP#1 used for? If CoreSight debug components are attached to it, then it too should have a ROM table.

Could you please run this command (on unmodified task version):\
`pyocd cmd -L pyocd.coresight.dap.trace=debug -c st`\
and attach the log? (The `st` command is just there to cause an exit.)

This command will report all the DP and AP register accesses and their values, so I can see what is causing the trouble.

There are two goals here: understand if pyocd has a bug, and see if there is a more general way to resolve this. Thanks for working with me on this! 😄

Reply:
Here is the log:

```
0005477:DEBUG:dap:write_dp:000001 (addr=0x00000000) = 0x0000001e
0005477:DEBUG:dap:write_dp:000003 (addr=0x00000008) = 0x00000000
0005477:DEBUG:dap:write_dp:000002 (addr=0x00000004) = 0x50000f00
0005481:DEBUG:dap:read_dp:000004 (addr=0x00000004) -> 0xf8000f40
0005481:DEBUG:dap:write_dp:000006 (addr=0x00000008) = 0x000000f0
0005483:DEBUG:dap:read_ap:000005 (addr=0x000000fc) -> 0x84770001
0005483:DEBUG:dap:write_dp:000008 (addr=0x00000008) = 0x010000f0
0005486:DEBUG:dap:read_ap:000007 (addr=0x010000fc) -> 0x54770002
0005486:DEBUG:dap:write_dp:000010 (addr=0x00000008) = 0x020000f0
0005489:DEBUG:dap:read_ap:000009 (addr=0x020000fc) -> 0x00000000
0005489:DEBUG:dap:write_dp:000012 (addr=0x00000008) = 0x030000f0
0005491:DEBUG:dap:read_ap:000011 (addr=0x030000fc) -> 0x00000000
0005491:DEBUG:dap:write_dp:000014 (addr=0x00000008) = 0x040000f0
0005493:DEBUG:dap:read_ap:000013 (addr=0x040000fc) -> 0x00000000
0005493:DEBUG:dap:write_dp:000016 (addr=0x00000008) = 0x000000f0
0005496:DEBUG:dap:read_ap:000015 (addr=0x000000fc) -> 0x84770001
0005496:DEBUG:dap:write_dp:000018 (addr=0x00000008) = 0x00000000
0005499:DEBUG:dap:read_ap:000017 (addr=0x00000000) -> 0x03800052
0005499:DEBUG:dap:write_dp:000020 (addr=0x00000008) = 0x000000f0
0005501:DEBUG:dap:read_ap:000019 (addr=0x000000f4) -> 0x00000000
0005501:DEBUG:dap:write_dp:000022 (addr=0x00000008) = 0x00000000
0005501:DEBUG:dap:write_ap:000021 (addr=0x00000000) = 0x4f800052
0005504:DEBUG:dap:read_ap:000023 (addr=0x00000000) -> 0x4f800052
0005504:DEBUG:dap:write_dp:000025 (addr=0x00000008) = 0x000000f0
0005506:DEBUG:dap:read_ap:000024 (addr=0x000000f8) -> 0xe00ff002
0005506:DEBUG:dap:write_dp:000027 (addr=0x00000008) = 0x00000000
0005506:DEBUG:dap:write_ap:000026 (addr=0x00000000) = 0x03800052
0005506:DEBUG:dap:write_dp:000029 (addr=0x00000008) = 0x010000f0
0005509:DEBUG:dap:read_ap:000028 (addr=0x010000fc) -> 0x54770002
0005509:DEBUG:dap:write_dp:000031 (addr=0x00000008) = 0x01000000
0005512:DEBUG:dap:read_ap:000030 (addr=0x01000000) -> 0x80000042
0005512:DEBUG:dap:write_dp:000033 (addr=0x00000008) = 0x010000f0
0005514:DEBUG:dap:read_ap:000032 (addr=0x010000f4) -> 0x00000000
0005514:DEBUG:dap:write_dp:000035 (addr=0x00000008) = 0x01000000
0005514:DEBUG:dap:write_ap:000034 (addr=0x01000000) = 0x80000041
0005514:DEBUG:dap:read_ap:000036 (addr=0x01000000) -> ...
0005517:DEBUG:dap:read_ap:000036 ...(addr=0x01000000) -> 0x80000042
0005517:DEBUG:dap:write_ap:000037 (addr=0x01000000) = 0xcf000042
0005519:DEBUG:dap:read_ap:000038 (addr=0x01000000) -> 0x80000042
0005519:DEBUG:dap:write_dp:000040 (addr=0x00000008) = 0x010000f0
0005521:DEBUG:dap:read_ap:000039 (addr=0x010000f8) -> 0xe00ff002
0005521:DEBUG:dap:write_dp:000042 (addr=0x00000008) = 0x01000000
0005521:DEBUG:dap:write_ap:000041 (addr=0x01000000) = 0x80000042
Exception while initing board: No cores were discovered!
Traceback (most recent call last):
  File "/usr/local/lib/python3.9/site-packages/pyocd/commands/commander.py", line 219, in _post_connect
    self.session.open(init_board=not self.args.no_init)
  File "/usr/local/lib/python3.9/site-packages/pyocd/core/session.py", line 459, in open
    self._board.init()
  File "/usr/local/lib/python3.9/site-packages/pyocd/board/board.py", line 99, in init
    self.target.init()
  File "/usr/local/lib/python3.9/site-packages/pyocd/core/soc_target.py", line 117, in init
    seq.invoke()
  File "/usr/local/lib/python3.9/site-packages/pyocd/utility/sequencer.py", line 208, in invoke
    resultSequence = call()
  File "/usr/local/lib/python3.9/site-packages/pyocd/coresight/coresight_target.py", line 244, in check_for_cores
    raise exceptions.DebugError("No cores were discovered!")
pyocd.core.exceptions.DebugError: No cores were discovered!
```

Reply:
By the way, this one ROM table on AP0, following is JLINK connect log:

```
Connecting to target via SWD
Found SW-DP with ID 0x6BA02477
DPIDR: 0x6BA02477
Scanning AP map to find all available APs
AP[2]: Stopped AP scan as end of AP map has been reached
AP[0]: AHB-AP (IDR: 0x84770001)
AP[1]: APB-AP (IDR: 0x54770002)
Iterating through AP map to find AHB-AP to use
AP[0]: Core found
AP[0]: AHB-AP ROM base: 0xE00FF000
CPUID register: 0x410FD214. Implementer code: 0x41 (ARM)
Found Cortex-M33 r0p4, Little endian.
FPUnit: 8 code (BP) slots and 0 literal slots
Security extension: not implemented
CoreSight components:
ROMTbl[0] @ E00FF000
ROMTbl[0][0]: E000E000, CID: B105900D, PID: 000BBD21 Cortex-M33
ROMTbl[0][1]: E0001000, CID: B105900D, PID: 000BBD21 DWT
ROMTbl[0][2]: E0002000, CID: B105900D, PID: 000BBD21 FPB
ROMTbl[0][3]: E0000000, CID: B105900D, PID: 000BBD21 ITM
ROMTbl[0][5]: E0041000, CID: B105900D, PID: 002BBD21 ETM
ROMTbl[0][6]: E0042000, CID: B105900D, PID: 000BBD21 CSS600-CTI
Cortex-M33 identified.
```

Reply:
Thanks for the logs! That explains it. The key lines are these reads of the `BASE` register at offset 0xf8 from the first two APs 0 and 1:

```
0005506:DEBUG:dap:read_ap:000024 (addr=0x000000f8) -> 0xe00ff002
```

```
0005521:DEBUG:dap:read_ap:000039 (addr=0x010000f8) -> 0xe00ff002
```


This is the layout of the MEM-AP `BASE` register (from Arm ADI v5.2 specification, debug_interface_v5_2_architecture_specification_IHI0031F.pdf, §C2.6.1, page C2-172):

<img width="775" alt="image" src="https://user-images.githubusercontent.com/3116536/145625847-ca13f597-4f3c-4b5c-bed0-971297293b92.png">

The important parts of the returned value 0xe00ff002 is the two least significant bits = 0x2:

- Format, bit[1]: 0x1, meaning the value is the current (non-deprecated) base register format. All good.
- P, bit[0]: 0x0, Not set! _**This is the problem. P=0 means that there is no ROM table.**_

From the same ADI v5.2 specification:

<img width="633" alt="image" src="https://user-images.githubusercontent.com/3116536/145627030-4ba82ec9-3469-42e6-9539-503daa3c58fa.png">

So, pyocd isn't doing anything wrong. Somehow the chip has an incorrectly configured `BASE` register.

One question is why the J-Link continues to read the ROM table even though the P bit is cleared to say "there is no ROM table here". Maybe because it sees that the remainder of the `BASE` value is non-zero?

Does the J-Link have specific support for the YTM32B1MD0? If so, it's probably just using a workaround they added.

----

Now the question is how to best work around this. It is important for pyocd to parse the ROM table, so that all the correct CoreSight components are detects. For instance, breakpoints and watchpoints will not work if the FPB and DWT components are not detected in the ROM table and hooked up to the core.

I'll think about this and reply shortly with a proposed method to work around the invalid P bit.

Reply:
Thanks Chris, Yes this is the root cause, I also got invalid error when using ST-LINK/CMSIS-DAP debugger on KEIL IDE, as this is a hardware issue, we can't fix this on chip, so we are also working on this issue. Please help me if you have a work around, thanks.

Reply:
For Keil MDK, you'll need to create a [CMSIS Device Family Pack](https://open-cmsis-pack.github.io/Open-CMSIS-Pack-Spec). The CMSIS-Pack specification allows for a [datapatch element](https://open-cmsis-pack.github.io/Open-CMSIS-Pack-Spec/main/html/pdsc_family_pg.html#element_dbg_datapatch) to be defined to override the value returned by the hardware from an AP or memory read.

In fact, the example included in the documentation is exactly what you need:

```xml
<debug>
    <!-- Patch the ROM Table base address -->
    <datapatch  type=”AP”__dp="0" __ap="0" address="0xF8" value="0xE00FF003" info="AP BASE Register, ROM Table at 0xE00FF000"/>
</debug>
```

While pyocd supports CMSIS Packs, it doesn't have a data patch mechanism. Adding patch support would be a good way to work around the issue, it's more work than I have time for right now.

A good mid-term work around will be to manually fix-up the AP's ROM table base address value after it's read. This is done by wrapping the init call sequence tasks similarly to what you've done, but more focused. I'll reply in a little bit with the code to insert.

Reply:
Thanks a lot, It works on KEIL. This really help us a lot.

Reply:
Hi @majorlin, my apologies that it took a couple weeks to get back to you! (Got distracted with holidays and lots of pyocd bug fixing…) I'm glad that Keil is working for you now!

This function below is a minimal change to correct the ROM table address on AP#0.

```py
    def create_init_sequence(self):
        # Insert init task to correct the ROM table base address value that incorrectly has the
        # P (preset) bit 0 cleared in hardware.

        def fixup_rom_base():
            self.aps[0].rom_addr = 0xE00FF00
            self.aps[0].has_rom_table = True
        
        seq = super().create_init_sequence()
        seq.wrap_task('discovery',
            lambda seq: seq.insert_after('create_aps', ('fixup_rom_base', fixup_rom_base))
            )
        return seq
```

Regarding APB-AP#1: Presumably this AP on your chip has some debug components attached to it, probably trace related. That's a pretty common configuration for Cortex-M SoCs.

What components are attached (within in the memory space accessible via APB-AP#1, which may or may not be the system memory space)?
Is there a ROM table?

If AP#1 does have a ROM table, we can extend the `fixup_rom_base()` function above to correct the ROM table address for AP#1 too.

Reply:
Hi Chris, Have not login GitHub for a long time... Thanks for the solution, I just tried it on our chips, it works as expected!! I just updated the flash loader and renamed device to me0.